src/pool.c

```diff
@@ -1935,30 +1936,42 @@ rpc_on_wallet_transferred(const char* data, rpc_callback_t *callback)
         goto cleanup;
     }
     payment_t *payment = (payment_t*) callback->data;
-    for (; payment->amount; payment++)
+    if (payment)
     {
         MDB_cursor_op op = MDB_SET;
         MDB_val key = {ADDRESS_MAX, (void*)payment->address};
         MDB_val val;
+
+        if (error)
+        {
+            log_warn("Error seen on transfer for %s with amount %"PRIu64,
+                    payment->address, payment->amount);
+            payment->error = 1;
+            goto skip_balance;
+        }
+
         rc = mdb_cursor_get(cursor, &key, &val, op);
         if (rc == MDB_NOTFOUND)
         {
             log_error("Payment made to non-existent address");
-            continue;
+            goto skip_balance;
         }
         else if (rc != 0 && rc != MDB_NOTFOUND)
         {
             err = mdb_strerror(rc);
             log_error("%s", err);
-            continue;
+            goto skip_balance;
         }
         uint64_t current_amount = *(uint64_t*)val.mv_data;
-        current_amount -= payment->amount;
-        if (error)
-        {
-            log_warn("Error seen on transfer for %s with amount %"PRIu64,
-                    payment->address, payment->amount);
+
+        if (current_amount >= payment->amount) {
+            current_amount -= payment->amount;
+        } else {
+            log_error("Payment was more than balance: %"PRIu64" > %"PRIu64,
+                      payment->amount, current_amount);
+            current_amount = 0;
         }
```

Comment:
Can you PR this separately please as this can be applied without changing the RPC/payout semantics.

Reply:
sure